src/lsm/k_way_merge.zig

```diff
@@ -24,207 +29,260 @@ pub fn KWayMergeIteratorType(
     ) error{ Empty, Drained }!Key,
     comptime stream_pop: fn (context: *Context, stream_index: u32) Value,
     /// Returns true if stream A has higher precedence than stream B.
-    /// This is used to deduplicate values across streams.
+    /// This is used to break ties across streams.
     comptime stream_precedence: fn (context: *const Context, a: u32, b: u32) bool,
 ) type {
-    return struct {
-        const KWayMergeIterator = @This();
+    comptime assert(options.streams_max >= 1);
 
-        context: *Context,
-        streams_count: u32,
-        direction: Direction,
-        state: enum { loading, iterating },
+    return struct {
 
-        /// Array of keys, with each key representing the next key in each stream.
-        ///
-        /// `keys` is *almost* structured as a binary heap — to become a heap, streams[0] must be
-        /// peeked and sifted (see pop_internal()).
-        ///
-        /// * When `direction=ascending`, keys are ordered low-to-high.
-        /// * When `direction=descending`, keys are ordered high-to-low.
-        /// * Equivalent keys are ordered from high precedence to low.
-        keys: [streams_max]Key = undefined,
+        // K-way merge via a loser tree algorithm (Knuth Volume 3 p. 253).
+        // Merges k sorted streams using a tournament (loser) tree.
+        // The current global winner lives in `contender`. Internal nodes in `losers`
+        // store the losers of the last comparisons along the root-to-leaf paths.
+        //     0 (winner, contender)
+        //
+        //     1
+        //    / \
+        //   2   3
+        //  / \ / \
+        // 4  5 6  7
+        // -------------
+        // K input streams
+        //
+        // The internal nodes are organized in a flat Eytzinger layout.
+        // That is the tree above is stored as [1][2][3][4][5][6][7].
+        //
+        // There are also a few optimizations that seem to be helpful, but did not work, such as:
+        // - Only store stream_id in the inner nodes, and have a heads array for the first keys
+        //   of the streams. This reduces space (densely packed) and the bytes in swap.
+        //   Unfortunately, it made the code much slower.
```

Comment:
Please, promote this cool explanation to a [top-level comment](https://ziglang.org/documentation/0.14.1/#Top-Level-Doc-Comments).